In [0]:
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("credit_risk_eda").getOrCreate()
spark

In [0]:
application_df=spark.table("workspace.credit_risk_data_delta.application_train")
application_df.printSchema()

root
 |-- SK_ID_CURR: integer (nullable = true)
 |-- TARGET: integer (nullable = true)
 |-- NAME_CONTRACT_TYPE: string (nullable = true)
 |-- CODE_GENDER: string (nullable = true)
 |-- FLAG_OWN_CAR: string (nullable = true)
 |-- FLAG_OWN_REALTY: string (nullable = true)
 |-- CNT_CHILDREN: integer (nullable = true)
 |-- AMT_INCOME_TOTAL: double (nullable = true)
 |-- AMT_CREDIT: double (nullable = true)
 |-- AMT_ANNUITY: double (nullable = true)
 |-- AMT_GOODS_PRICE: double (nullable = true)
 |-- NAME_TYPE_SUITE: string (nullable = true)
 |-- NAME_INCOME_TYPE: string (nullable = true)
 |-- NAME_EDUCATION_TYPE: string (nullable = true)
 |-- NAME_FAMILY_STATUS: string (nullable = true)
 |-- NAME_HOUSING_TYPE: string (nullable = true)
 |-- REGION_POPULATION_RELATIVE: double (nullable = true)
 |-- DAYS_BIRTH: integer (nullable = true)
 |-- DAYS_EMPLOYED: integer (nullable = true)
 |-- DAYS_REGISTRATION: double (nullable = true)
 |-- DAYS_ID_PUBLISH: integer (nullable = true)
 |-- OWN_CAR_AG

In [0]:
display(application_df.limit(10))

SK_ID_CURR TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY CNT_CHILDREN AMT_INCOME_TOTAL AMT_CREDIT AMT_ANNUITY AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE NAME_EDUCATION_TYPE NAME_FAMILY_STATUS NAME_HOUSING_TYPE REGION_POPULATION_RELATIVE DAYS_BIRTH DAYS_EMPLOYED DAYS_REGISTRATION DAYS_ID_PUBLISH OWN_CAR_AGE FLAG_MOBIL FLAG_EMP_PHONE FLAG_WORK_PHONE FLAG_CONT_MOBILE FLAG_PHONE FLAG_EMAIL OCCUPATION_TYPE CNT_FAM_MEMBERS REGION_RATING_CLIENT REGION_RATING_CLIENT_W_CITY WEEKDAY_APPR_PROCESS_START HOUR_APPR_PROCESS_START REG_REGION_NOT_LIVE_REGION REG_REGION_NOT_WORK_REGION LIVE_REGION_NOT_WORK_REGION REG_CITY_NOT_LIVE_CITY REG_CITY_NOT_WORK_CITY LIVE_CITY_NOT_WORK_CITY ORGANIZATION_TYPE EXT_SOURCE_1 EXT_SOURCE_2 EXT_SOURCE_3 APARTMENTS_AVG BASEMENTAREA_AVG YEARS_BEGINEXPLUATATION_AVG YEARS_BUILD_AVG COMMONAREA_AVG ELEVATORS_AVG ENTRANCES_AVG FLOORSMAX_AVG FLOORSMIN_AVG LANDAREA_AVG LIVINGAPARTMENTS_AVG LIVINGAREA_AVG NONLIVINGAPARTMENTS_AVG NONLIVINGAREA_AVG APARTMENTS_MODE BASEMENTAREA_MODE YEARS_BEGINEXPLUATATION_MODE YEARS_BUILD_MODE COMMONAREA_MODE ELEVATORS_MODE ENTRANCES_MODE FLOORSMAX_MODE FLOORSMIN_MODE LANDAREA_MODE LIVINGAPARTMENTS_MODE LIVINGAREA_MODE NONLIVINGAPARTMENTS_MODE NONLIVINGAREA_MODE APARTMENTS_MEDI BASEMENTAREA_MEDI YEARS_BEGINEXPLUATATION_MEDI YEARS_BUILD_MEDI COMMONAREA_MEDI ELEVATORS_MEDI ENTRANCES_MEDI FLOORSMAX_MEDI FLOORSMIN_MEDI LANDAREA_MEDI LIVINGAPARTMENTS_MEDI LIVINGAREA_MEDI NONLIVINGAPARTMENTS_MEDI NONLIVINGAREA_MEDI FONDKAPREMONT_MODE HOUSETYPE_MODE TOTALAREA_MODE WALLSMATERIAL_MODE EMERGENCYSTATE_MODE OBS_30_CNT_SOCIAL_CIRCLE DEF_30_CNT_SOCIAL_CIRCLE OBS_60_CNT_SOCIAL_CIRCLE DEF_60_CNT_SOCIAL_CIRCLE DAYS_LAST_PHONE_CHANGE FLAG_DOCUMENT_2 FLAG_DOCUMENT_3 FLAG_DOCUMENT_4 FLAG_DOCUMENT_5 FLAG_DOCUMENT_6 FLAG_DOCUMENT_7 FLAG_DOCUMENT_8 FLAG_DOCUMENT_9 FLAG_DOCUMENT_10 FLAG_DOCUMENT_11 FLAG_DOCUMENT_12 FLAG_DOCUMENT_13 FLAG_DOCUMENT_14 FLAG_DOCUMENT_15 FLAG_DOCUMENT_16 FLAG_DOCUMENT_17 FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY AMT_REQ_CREDIT_BUREAU_WEEK AMT_REQ_CREDIT_BUREAU_MON AMT_REQ_CREDIT_BUREAU_QRT AMT_REQ_CREDIT_BUREAU_YEAR 100002 1 Cash loans M N Y 0 202500.0 406597.5 24700.5 351000.0 Unaccompanied Working Secondary / secondary special Single / not married House / apartment 0.018801 -9461 -637 -3648.0 -2120 null 1 1 0 1 1 0 Laborers 1.0 2 2 WEDNESDAY 10 0 0 0 0 0 0 Business Entity Type 3 0.08303696739132256 0.2629485927471776 0.13937578009978951 0.0247 0.0369 0.9722 0.6192 0.0143 0.0 0.069 0.0833 0.125 0.0369 0.0202 0.019 0.0 0.0 0.0252 0.0383 0.9722 0.6341 0.0144 0.0 0.069 0.0833 0.125 0.0377 0.022 0.0198 0.0 0.0 0.025 0.0369 0.9722 0.6243 0.0144 0.0 0.069 0.0833 0.125 0.0375 0.0205 0.0193 0.0 0.0 reg oper account block of flats 0.0149 Stone, brick No 2.0 2.0 2.0 2.0 -1134.0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0 0.0 0.0 0.0 0.0 1.0 100003 0 Cash loans F N N 0 270000.0 1293502.5 35698.5 1129500.0 Family State servant Higher education Married House / apartment 0.003540999999999999 -16765 -1188 -1186.0 -291 null 1 1 0 1 1 0 Core staff 2.0 1 1 MONDAY 11 0 0 0 0 0 0 School 0.3112673113812225 0.6222457752555098 null 0.0959 0.0529 0.9851 0.7959999999999999 0.0605 0.08 0.0345 0.2917 0.3333 0.013 0.0773 0.0549 0.0039 0.0098 0.0924 0.0538 0.9851 0.804 0.0497 0.0806 0.0345 0.2917 0.3333 0.0128 0.079 0.0554 0.0 0.0 0.0968 0.0529 0.9851 0.7987 0.0608 0.08 0.0345 0.2917 0.3333 0.0132 0.0787 0.0558 0.0039 0.01 reg oper account block of flats 0.0714 Block No 1.0 0.0 1.0 0.0 -828.0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0 0.0 0.0 0.0 0.0 0.0 100004 0 Revolving loans M Y Y 0 67500.0 135000.0 6750.0 135000.0 Unaccompanied Working Secondary / secondary special Single / not married House / apartment 0.010032 -19046 -225 -4260.0 -2531 26.0 1 1 1 1 1 0 Laborers 1.0 2 2 MONDAY 9 0 0 0 0 0 0 Government null 0.5559120833904428 0.7295666907060153 null null null null null null null null null null null nul

In [0]:
display(application_df.describe())

summary SK_ID_CURR TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY CNT_CHILDREN AMT_INCOME_TOTAL AMT_CREDIT AMT_ANNUITY AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE NAME_EDUCATION_TYPE NAME_FAMILY_STATUS NAME_HOUSING_TYPE REGION_POPULATION_RELATIVE DAYS_BIRTH DAYS_EMPLOYED DAYS_REGISTRATION DAYS_ID_PUBLISH OWN_CAR_AGE FLAG_MOBIL FLAG_EMP_PHONE FLAG_WORK_PHONE FLAG_CONT_MOBILE FLAG_PHONE FLAG_EMAIL OCCUPATION_TYPE CNT_FAM_MEMBERS REGION_RATING_CLIENT REGION_RATING_CLIENT_W_CITY WEEKDAY_APPR_PROCESS_START HOUR_APPR_PROCESS_START REG_REGION_NOT_LIVE_REGION REG_REGION_NOT_WORK_REGION LIVE_REGION_NOT_WORK_REGION REG_CITY_NOT_LIVE_CITY REG_CITY_NOT_WORK_CITY LIVE_CITY_NOT_WORK_CITY ORGANIZATION_TYPE EXT_SOURCE_1 EXT_SOURCE_2 EXT_SOURCE_3 APARTMENTS_AVG BASEMENTAREA_AVG YEARS_BEGINEXPLUATATION_AVG YEARS_BUILD_AVG COMMONAREA_AVG ELEVATORS_AVG ENTRANCES_AVG FLOORSMAX_AVG FLOORSMIN_AVG LANDAREA_AVG LIVINGAPARTMENTS_AVG LIVINGAREA_AVG NONLIVINGAPARTMENTS_AVG NONLIVINGAREA_AVG APARTMENTS_MODE BASEMENTAREA_MODE YEARS_BEGINEXPLUATATION_MODE YEARS_BUILD_MODE COMMONAREA_MODE ELEVATORS_MODE ENTRANCES_MODE FLOORSMAX_MODE FLOORSMIN_MODE LANDAREA_MODE LIVINGAPARTMENTS_MODE LIVINGAREA_MODE NONLIVINGAPARTMENTS_MODE NONLIVINGAREA_MODE APARTMENTS_MEDI BASEMENTAREA_MEDI YEARS_BEGINEXPLUATATION_MEDI YEARS_BUILD_MEDI COMMONAREA_MEDI ELEVATORS_MEDI ENTRANCES_MEDI FLOORSMAX_MEDI FLOORSMIN_MEDI LANDAREA_MEDI LIVINGAPARTMENTS_MEDI LIVINGAREA_MEDI NONLIVINGAPARTMENTS_MEDI NONLIVINGAREA_MEDI FONDKAPREMONT_MODE HOUSETYPE_MODE TOTALAREA_MODE WALLSMATERIAL_MODE EMERGENCYSTATE_MODE OBS_30_CNT_SOCIAL_CIRCLE DEF_30_CNT_SOCIAL_CIRCLE OBS_60_CNT_SOCIAL_CIRCLE DEF_60_CNT_SOCIAL_CIRCLE DAYS_LAST_PHONE_CHANGE FLAG_DOCUMENT_2 FLAG_DOCUMENT_3 FLAG_DOCUMENT_4 FLAG_DOCUMENT_5 FLAG_DOCUMENT_6 FLAG_DOCUMENT_7 FLAG_DOCUMENT_8 FLAG_DOCUMENT_9 FLAG_DOCUMENT_10 FLAG_DOCUMENT_11 FLAG_DOCUMENT_12 FLAG_DOCUMENT_13 FLAG_DOCUMENT_14 FLAG_DOCUMENT_15 FLAG_DOCUMENT_16 FLAG_DOCUMENT_17 FLAG_DOCUMENT_18 FLAG_DOCUMENT_19 FLAG_DOCUMENT_20 FLAG_DOCUMENT_21 AMT_REQ_CREDIT_BUREAU_HOUR AMT_REQ_CREDIT_BUREAU_DAY AMT_REQ_CREDIT_BUREAU_WEEK AMT_REQ_CREDIT_BUREAU_MON AMT_REQ_CREDIT_BUREAU_QRT AMT_REQ_CREDIT_BUREAU_YEAR count 307511 307511 307511 307511 307511 307511 307511 307511 307511 307499 307233 306219 307511 307511 307511 307511 307511 307511 307511 307511 307511 104582 307511 307511 307511 307511 307511 307511 211120 307509 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 134133 306851 246546 151450 127568 157504 103023 92646 143620 152683 154491 98869 124921 97312 153161 93997 137829 151450 127568 157504 103023 92646 143620 152683 154491 98869 124921 97312 153161 93997 137829 151450 127568 157504 103023 92646 143620 152683 154491 98869 124921 97312 153161 93997 137829 97216 153214 159080 151170 161756 306490 306490 306490 306490 307510 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 307511 265992 265992 265992 265992 265992 265992 mean 278180.51857657125 0.08072881945686496 null null null null 0.4170517477423572 168797.9192969845 599025.9997057016 27108.573909183444 538396.2074288895 null null null null null 0.02086811205778947 -16036.995066843137 63815.04590404896 -4986.120327538419 -2994.2023732484367 12.061090818687727 0.9999967480838083 0.8198893698111612 0.1993684778755882 0.9981334001060125 0.28106636835755466 0.0567199222141647 null 2.152665450442101 2.0524631639193394 2.031520823645333 null 12.063418869568894 0.015144173704355291 0.05076891558350758 0.04065870814377372 0.07817281333025486 0.23045354475124466 0.17955455252007246 null 0.5021298056566621 0.5143926741308463 0.5108529061800121 0.11744049917466033 0.08844221905180052 0.9777348581622732 0.752471432592686 0.04462071541135205 0.07894151232418305 0.1497246700681123 0.2262819070368362 0.23189350049050692 0.06633318417239835 0.10077477495068154 0.10739901933259607 0.008808672617213057 0.0283577

#### 1. Target Imbalance:
The `TARGET` variable has a mean of approximately **0.08**, confirming a **strong class imbalance** — only ~8% of applicants default.  
This requires:
- **Stratified sampling** during train-test split  
- Evaluation metrics like **AUC**, **F1**, or **log loss** over accuracy

---

#### 2. Missing Values:
Many columns — especially **apartment-level features** and **housing metadata** — have significantly fewer entries than 307,511, suggesting heavy missingness.  
Action Plan:
- Use null percentage summaries to decide on **imputation or removal**  
- Prioritize important columns for retention during feature engineering 
---

#### 3. Applicant Demographics & Behavior:

- **Age (`DAYS_BIRTH`)**: Mean age is ~44 years (16037 / 365)  
- **Employment (`DAYS_EMPLOYED`)**: Contains unrealistic values (e.g., 365243), which likely indicate retirees or placeholder codes — needs cleaning
- **Income & Credit**:  
  - `AMT_INCOME_TOTAL` and `AMT_CREDIT` show **high variance** and **extreme outliers**  
  - These features are ideal candidates for **log transformation** or **binning**

---

#### 4. Data Consistency & Encoding Clues:
Some categorical features such as `NAME_CONTRACT_TYPE`, `CODE_GENDER`, `OCCUPATION_TYPE`, etc., contain **ambiguous entries** like `'XNA'`.  
These will require:
- **Cleaning and consolidation**
- Proper encoding during (e.g., one-hot or target encoding)

---

#### 5. Engineered External Scores:
Features `EXT_SOURCE_1`, `EXT_SOURCE_2`, and `EXT_SOURCE_3`:
- Appear to be **pre-normalized** (values in 0–1 range)
- Typically **strong predictors** in credit risk modeling  
Must visualize their correlation with `TARGET` in `01_eda.ipynb`

---

#### 6. Document Flags and Count Features:
Flags like `FLAG_DOCUMENT_3`, `FLAG_PHONE`, etc.:
- Are mostly **binary indicators** or **low-range counts**
- Can be used directly or **combined into composite features** (e.g., total documents submitted)


In [0]:
# Calculating the number of Rows and Columns
num_rows = application_df.count()
num_cols = len(application_df.columns)
print(f"({num_rows}, {num_cols})")

(307511, 122)


In [0]:
from collections import Counter
Counter(dict(application_df.dtypes).values())

Counter({'double': 65, 'int': 41, 'string': 16})

In [0]:
from pyspark.sql.functions import col

# Filter only string (categorical) columns
categorical_cols = [field.name for field in application_df.schema.fields if field.dataType.simpleString() == 'string']

# Show number of unique values for each categorical column
for col_name in categorical_cols:
    count = application_df.select(col_name).distinct().count()
    print(f"{col_name}: {count}")

NAME_CONTRACT_TYPE: 2
CODE_GENDER: 3
FLAG_OWN_CAR: 2
FLAG_OWN_REALTY: 2
NAME_TYPE_SUITE: 8
NAME_INCOME_TYPE: 8
NAME_EDUCATION_TYPE: 5
NAME_FAMILY_STATUS: 6
NAME_HOUSING_TYPE: 6
OCCUPATION_TYPE: 19
WEEKDAY_APPR_PROCESS_START: 7
ORGANIZATION_TYPE: 58
FONDKAPREMONT_MODE: 5
HOUSETYPE_MODE: 4
WALLSMATERIAL_MODE: 8
EMERGENCYSTATE_MODE: 3
